compiler/main/PhaseTracker.cpp

```diff
@@ -27,117 +27,115 @@
   }
 };
+struct SortByMemory {
+  bool operator()(Pass const& a, Pass const& b) const {
+    return a.CompareByMemory(b);
+  }
+};
 
-static void PassesSortByTime(std::vector<Pass>& passes);
-
-static void PassesReport(const std::vector<Pass>& passes,
-                         unsigned long            totalTime);
+static void PassesReport(const std::vector<Pass>& passes, PhaseData total);
 
-static void PassesReport(FILE*                    fp,
-                         const std::vector<Pass>& passes,
-                         unsigned long            totalTime);
+static void
+PassesReport(FILE* fp, const std::vector<Pass>& passes, PhaseData total);
 
 /************************************* | **************************************
-*                                                                             *
-* Implementation of PhaseTracker                                              *
-*                                                                             *
-************************************** | *************************************/
+ *                                                                             *
+ * Implementation of PhaseTracker                                              *
+ *                                                                             *
+ ************************************** | *************************************/
 
-PhaseTracker::PhaseTracker()
-{
+PhaseTracker::PhaseTracker() {
   mPhaseId = 0;
 
   mTimer.start();
+  mMemoryTracker.start();
   StartPhase("startup");
 }
 
-PhaseTracker::~PhaseTracker()
-{
-  for (size_t i = 0; i < mPhases.size(); i++)
-    delete mPhases[i];
+PhaseTracker::~PhaseTracker() {
+  for (size_t i = 0; i < mPhases.size(); i++) delete mPhases[i];
 }
 
-void PhaseTracker::StartPhase(const char* name)
-{
+void PhaseTracker::StartPhase(const char* name) {
   StartPhase(name, 0, kPrimary);
 }
 
 void PhaseTracker::StartPhase(const char*            name,
-                              PhaseTracker::SubPhase subPhase)
-{
-  if (subPhase == kPrimary)
-    mPhaseId = mPhaseId + 1;
+                              PhaseTracker::SubPhase subPhase) {
+  if (subPhase == kPrimary) mPhaseId = mPhaseId + 1;
 
   StartPhase(name, mPhaseId, subPhase);
 }
 
-
-void PhaseTracker::StartPhase(const char* name,
-                              int         passId,
-                              SubPhase    subPhase)
-{
-  Phase* phase = new Phase(name, passId, subPhase, mTimer.elapsedUsecs());
+void PhaseTracker::StartPhase(const char* name, int passId, SubPhase subPhase) {
+  Phase* phase = new Phase(
+    name,
+    passId,
+    subPhase,
+    PhaseData(mTimer.elapsedUsecs(), mMemoryTracker.usedBytes())
+  );
 
   mPhases.push_back(phase);
 }
 
-void PhaseTracker::Stop()
-{
+void PhaseTracker::Stop() {
   mTimer.stop();
+  mMemoryTracker.stop();
 }
 
-void PhaseTracker::Resume()
-{
+void PhaseTracker::Resume() {
   mTimer.start();
+  mMemoryTracker.start();
```

Comment:
dunno if this matters much, but should `mTimer` be launched after `mMemoryTracker` to exclude the time it takes for memory tracker to compute usage?